src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs

```diff
@@ -1006,35 +1006,56 @@ public string Replace(char oldChar, char newChar)
             }
 
             // Copy the remaining characters, doing the replacement as we go.
-            ref ushort pSrc = ref Unsafe.Add(ref Unsafe.As<char, ushort>(ref _firstChar), copyLength);
-            ref ushort pDst = ref Unsafe.Add(ref Unsafe.As<char, ushort>(ref result._firstChar), copyLength);
+            ref ushort pSrc = ref Unsafe.Add(ref GetRawStringDataAsUInt16(), (uint)copyLength);
+            ref ushort pDst = ref Unsafe.Add(ref result.GetRawStringDataAsUInt16(), (uint)copyLength);
+            nuint i = 0;
 
-            if (Vector.IsHardwareAccelerated && remainingLength >= Vector<ushort>.Count)
+            if (Vector.IsHardwareAccelerated && Length >= Vector<ushort>.Count)
             {
-                Vector<ushort> oldChars = new Vector<ushort>(oldChar);
-                Vector<ushort> newChars = new Vector<ushort>(newChar);
+                Vector<ushort> oldChars = new(oldChar);
+                Vector<ushort> newChars = new(newChar);
 
-                do
+                Vector<ushort> original;
+                Vector<ushort> equals;
+                Vector<ushort> results;
+
+                nuint lengthToExamine = remainingLength - (nuint)Vector<ushort>.Count;
+
+                if ((nint)lengthToExamine > 0)
                 {
```

Comment:
Nitpick: Correct me if I'm wrong, but it seems like you are allowing an underflow on `1022` and then reinterpreting the result as a signed number on `1024` to account for it. Would it be cleaner to cast `remainingLength` to `nint` before the subtraction, and then have `lengthToExample` be an `nint`?

Reply:
The underflow and cast to `nint` is intentional here, as it safes some casts to `nuint` in the rest of the code.
So one cast here, instead of 4 (or more) in the remaining code. Just a style thing. Could be written with `nint` too.

Reply:
That sounds reasonable to me. I could see something like the following being clearer, but I'm on the fence. I'll leave it up to you.

```suggestion
                if (remainingLength > (nuint)Vector<ushort>.Count)
                {
                    nuint lengthToExamine = remainingLength - (nuint)Vector<ushort>.Count;
                    
```

Reply:
Yep, that's nicer to read, so I took that suggestion, thanks!

Only micro downside is that there's a `cmp` instead of a `test` now in machine code (comparison to 0 emits `test`), and test can be fused at the cpu-level. But given that the code inside the branch will be dominant, this won't have any real effect.